exporters/otlp/common/src/main/java/io/opentelemetry/exporter/otlp/internal/okhttp/OkHttpRetrier.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.exporter.otlp.internal.okhttp;
+
+import io.opentelemetry.exporter.otlp.internal.RetryPolicy;
+import java.io.IOException;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import okhttp3.Interceptor;
+import okhttp3.Response;
+
+/** Retrier of OkHttp requests. */
+public final class OkHttpRetrier implements Interceptor {
+
+  private final RetryPolicy retryPolicy;
+  private final Function<Response, Boolean> isRetriableError;
+  private final Sleeper sleeper;
+  private final BoundedLongGenerator randomLong;
+
+  /** Constructs a new retrier. */
+  public OkHttpRetrier(RetryPolicy retryPolicy, Function<Response, Boolean> isRetriableError) {
+    this(
+        retryPolicy,
+        isRetriableError,
+        TimeUnit.NANOSECONDS::sleep,
+        bound -> ThreadLocalRandom.current().nextLong(bound));
+  }
+
+  // Visible for testing
+  OkHttpRetrier(
+      RetryPolicy retryPolicy,
+      Function<Response, Boolean> isRetriableError,
+      Sleeper sleeper,
+      BoundedLongGenerator randomLong) {
+    this.retryPolicy = retryPolicy;
+    this.isRetriableError = isRetriableError;
+    this.sleeper = sleeper;
+    this.randomLong = randomLong;
+  }
+
+  @Override
+  public Response intercept(Chain chain) throws IOException {
+    Response response = chain.proceed(chain.request());
+    if (!Boolean.TRUE.equals(isRetriableError.apply(response))) {
+      return response;
+    }
+    long nextBackoffNanos = retryPolicy.getInitialBackoff().toNanos();
+    // Made an attempt already, starting with the 2nd.
+    for (int attempt = 2; attempt <= retryPolicy.getMaxAttempts(); attempt++) {
+      // https://github.com/grpc/proposal/blob/master/A6-client-retries.md#exponential-backoff
+      // Careful - the document says "retry attempt" but our maxAttempts includes the original
+      // request. The numbers are tricky but the unit test should be more clear.
+      long upperBoundNanos = Math.min(nextBackoffNanos, retryPolicy.getMaxBackoff().toNanos());
+      long backoffNanos = randomLong.get(upperBoundNanos);
```

Comment:
I think this logic is correct, but IMO it reads cleaner if its extracted into a helper function. Here's what I have locally for a `http/protobuf` retry interceptor:
```
private static int backoffMillisWithJitter(int attempt, RetryPolicy retryPolicy) {
    return RANDOM.nextInt(backoffMillisTarget(attempt, retryPolicy));
  }

  // Visible for testing
  static int backoffMillisTarget(int attempt, RetryPolicy retryPolicy) {
    long initialBackoffMillis = retryPolicy.getInitialBackoff().toMillis();
    if (attempt == 1) {
      return (int) initialBackoffMillis;
    }
    long target =
        (long) (initialBackoffMillis * Math.pow(retryPolicy.getBackoffMultiplier(), attempt));
    return (int) Math.min(target, retryPolicy.getMaxBackoff().toMillis());
  }
```
Would likely need to be extracted to a helper function once the `http/protobuf` retry stuff lands anyway.

Reply:
Sorry for writing this while you also had code in the work! I was thinking that the only difference between the two would be the definition if `isRetryable` but otherwise the same intercceptor could be reused (I don't think it hurts to have support for retry-after even in the one used by gRPC). See anything that would prevent that?

Reply:
Also I had a similar pattern but changed to this one to avoid redundant math in Math.pow, similar to what gRPC code was doing. I don't think it's a huge deal but would keep it for that reason if it is possible to reuse the interceptor instead of just a helper method.

Reply:
No problem! I do think this interceptor can be reused. 

One note: I believe that if `chain.proceed(Request)` is called and throws an `IOException` that we should retry, rather than letting it propagate to `Callback#onFailure(Call, IOException)`.

Reply:
Unfortunately it will throw IOException for things like a bad TLS certificate which we shouldn't retry on. Some smart checking of `instanceof` on the exception might work but I'm not sure the best way to implement that right now.

Reply:
Oh yeah, OkHttp has automatic retry on connection failure so I guess we can generally live without retrying on IOException? https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/-builder/retry-on-connection-failure/

Reply:
Yeah, probably fine ignoring IOException.